crates/saya/core/src/prover/program_input.rs

```diff
@@ -0,0 +1,192 @@
+use katana_primitives::contract::ContractAddress;
+use katana_primitives::state::StateUpdates;
+use katana_primitives::trace::{EntryPointType, TxExecInfo};
+use starknet::core::types::FieldElement;
+
+use super::state_diff::state_updates_to_json_like;
+
+/// Based on https://github.com/cartridge-gg/piltover/blob/2be9d46f00c9c71e2217ab74341f77b09f034c81/src/snos_output.cairo#L19-L20
+/// With the new state root computed by ten prover.
+pub struct ProgramInput {
+    pub prev_state_root: FieldElement,
+    pub block_number: FieldElement,
+    pub block_hash: FieldElement,
+    pub config_hash: FieldElement,
+    pub message_to_starknet_segment: Vec<MessageToStarknet>,
+    pub message_to_appchain_segment: Vec<MessageToAppchain>,
+    pub state_updates: StateUpdates,
+}
+
+pub fn extract_messages(
+    exec_infos: &Vec<TxExecInfo>,
+) -> (Vec<MessageToStarknet>, Vec<MessageToAppchain>) {
+    let message_to_starknet_segment = exec_infos
+        .iter()
+        .map(|t| t.execute_call_info.iter().chain(t.validate_call_info.iter())) // Take into account both validate and execute calls.
+        .flatten()
+        .map(|c| { // Flatten the recursive call structure.
+            let mut to_visit = vec![c];
+            let mut all = vec![c];
+
+            while let Some(c) = to_visit.pop() {
+                to_visit.extend(c.inner_calls.iter().rev());
+                all.extend(c.inner_calls.iter().rev());
+            }
+            all
+        })
+        .flatten()
+        .map(|c| c.l2_to_l1_messages.iter()) // take all messages
+        .flatten()
+        .map(|m| MessageToStarknet { // Parse them to the format understood by the prover.
+            from_address: m.from_address,
+            to_address: ContractAddress::from(m.to_address),
+            payload: m.payload.clone(),
+        })
+        .collect();
+
+    let message_to_appchain_segment = exec_infos
+        .iter()
+        .map(|t| t.execute_call_info.iter())
+        .flatten()
+        .filter(|c| c.entry_point_type == EntryPointType::L1Handler)
+        .map(|c| MessageToAppchain {
+            from_address: c.caller_address,
+            to_address: c.contract_address,
+            nonce: FieldElement::from(0u64), // TODO: extract nonce
```

Comment:
To have the nonce we may want to parse the transactions instead of the transaction outputs and the nonce should be present here.